File: SafetyProgram.Base/CommandInvoker.cs
using System;
using System.Collections.Generic;
using SafetyProgram.Base.Interfaces;

namespace SafetyProgram.Base
{
    /// <summary>
    /// Defines an ICommandInvoker. A class that will execute and buffer IInvokedCommands.
    /// </summary>
    public sealed class CommandInvoker : ICommandInvoker
    {
        private readonly Stack<IInvokedCommand> futureCommands;
        private readonly Stack<IInvokedCommand> pastCommands;

        public CommandInvoker()
        {
            futureCommands = new Stack<IInvokedCommand>();
            pastCommands = new Stack<IInvokedCommand>();
        }

        public void InvokeCommand(IInvokedCommand command)
        {
            //Execute a new command:
            //  Clear future commands (this command represents a new 'branch' in the undo-redo chain)
            //  Execute the command
            //  Add the command into the past
            invoker(() =>
            {
                futureCommands.Clear();
                command.Execute();
                pastCommands.Push(command);
            });
        }

        public void Undo()
        {
            //Undo a command:
            //  Pop a command off the 'past' stack.
            //  Unexecute that command.
            //  Put the unexecuted command onto the 'future' stack.
            invoker(() =>
            {
                var command = pastCommands.Pop();
                command.UnExecute();
                futureCommands.Push(command);
            });
        }

        public bool CanUndo()
        {
            return (pastCommands.Count > 0) ? true : false;
        }

        public event EventHandler<GenericPropertyChangedEventArg<bool>> CanUndoChanged;

        public void Redo()
        {
            //Redo a command
            //  Pop a command off the 'future' stack.
            //  Execute the command.
            //  Put the executed command into the 'past' stack.
            invoker(() =>
            {
                var command = futureCommands.Pop();
                command.Execute();
                pastCommands.Push(command);
            });
        }

        public bool CanRedo()
        {
            return (futureCommands.Count > 0) ? true : false;
        }      

        public event EventHandler<GenericPropertyChangedEventArg<bool>> CanRedoChanged;

        /// <summary>
        /// Invokes actions. Checks if the CanUndo and CanRedo states changed during invokation
        /// </summary>
        /// <param name="invokedAction"></param>
        private void invoker(Action invokedAction)
        {
            //Invokes an action, checking if the CanUndo and CanRedo states change after invokation.

            //Store pre-invocation CanUndo/CanRedo states
            bool canUndoBefore = CanUndo();
            bool canRedoBefore = CanRedo();

            //Invoke the action
            invokedAction();

            //Check if the states have changed. If they have, trigger the necessay events.
            {
                bool canUndoAfter = CanUndo();
                if (canUndoAfter != canUndoBefore)
                {
                    CanUndoChanged.Raise(this, canUndoAfter);
                }
            }
            {
                bool canRedoAfter = CanRedo();
                if (canRedoAfter != canRedoBefore)
                {
                    CanRedoChanged.Raise(this, canRedoAfter);
                }
            }
        }
    }
}

File: SafetyProgram.Base/CommandInvoker.fs
﻿namespace SafetyProgram.Base

open System.Collections.Generic

type ReversibleCommand = {
    Execute : unit -> unit
    UnExecute : unit -> unit
}

type ICommandInvoker = 
    abstract invoke : ReversibleCommand -> unit

type ICommandController =
    
    inherit ICommandInvoker

    abstract Undo : unit -> unit
    abstract CanUndo : unit -> bool
    abstract CanUndoChanged : IEvent<bool>

    abstract Redo : unit -> unit
    abstract CanRedo : unit -> bool
    abstract CanRedoChanged : IEvent<bool>

type CommandController () =

    let canUndoChanged = new Event<_>()
    let canRedoChanged = new Event<_>()

    // Hold commands
    let futureCommands = new Stack<_>()
    let pastCommands = new Stack<_>()

    // Predicates for executing a past or future command.
    let canUndo() = pastCommands.Count > 0
    let canRedo() = futureCommands.Count > 0

    // Invoke a function (not a command)
    let invoke f = 
        let canUndoBefore = canUndo()
        let canRedoBefore = canRedo()

        f()

        // Echo to public event that a change in state has occured.
        let canUndoAfter = canUndo()
        if canUndoAfter <> canUndoBefore then
            canUndoChanged.Trigger canUndoAfter

        // Same as above.
        let canRedoAfter = canRedo()
        if canRedoAfter <> canRedoBefore then
            canRedoChanged.Trigger canRedoAfter

    // Outer interface declaration.
    interface ICommandController with
        // Invoke a command.
        member this.invoke (cmd) =

            let f() = 
                futureCommands.Clear()
                cmd.Execute()
                pastCommands.Push cmd

            invoke f

        // Undo the last invoked command.
        member this.Undo() = 
            
            let f() = 
                let cmd = pastCommands.Pop ()
                cmd.UnExecute ()
                futureCommands.Push cmd

            invoke f

        member this.CanUndo() = canUndo()

        member this.CanUndoChanged = canUndoChanged.Publish

        // Redo the last undone command.
        member this.Redo() = 
            
            let f() = 
                let cmd = futureCommands.Pop()
                cmd.Execute()
                pastCommands.Push cmd

            invoke f

        member this.CanRedo() = canRedo()

        member this.CanRedoChanged = canRedoChanged.Publish